runtime/js-compute-runtime/builtins/cache-core.cpp

```diff
@@ -0,0 +1,1162 @@
+                                                            JS::HandleValue options_val) {
+  host_api::CacheLookupOptions options;
+  if (!options_val.isUndefined()) {
+    if (!options_val.isObject()) {
+      JS_ReportErrorASCII(cx, "options argument must be an object");
+      return JS::Result<host_api::CacheLookupOptions>(JS::Error());
+    }
+    JS::RootedObject options_obj(cx, &options_val.toObject());
+    JS::RootedValue headers_val(cx);
+    if (!JS_GetProperty(cx, options_obj, "headers", &headers_val)) {
+      return JS::Result<host_api::CacheLookupOptions>(JS::Error());
+    }
+    // headers property is optional
+    if (!headers_val.isUndefined()) {
+      JS::RootedObject headersInstance(
+          cx,
+          JS_NewObjectWithGivenProto(cx, &builtins::Headers::class_, builtins::Headers::proto_obj));
+      if (!headersInstance) {
+        return JS::Result<host_api::CacheLookupOptions>(JS::Error());
+      }
+      auto headers = builtins::Headers::create(
+          cx, headersInstance, builtins::Headers::Mode::Standalone, nullptr, headers_val, true);
+      if (!headers) {
+        return JS::Result<host_api::CacheLookupOptions>(JS::Error());
+      }
+      JS::RootedValue headers_val(cx, JS::ObjectValue(*headers));
+      JS::RootedObject requestInstance(cx, Request::create_instance(cx));
+      if (!requestInstance) {
+        return JS::Result<host_api::CacheLookupOptions>(JS::Error());
+      }
+
+      JS::RootedValue input(cx, JS::StringValue(JS_NewStringCopyZ(cx, "http://example.com")));
+      JS::RootedObject request(cx,
+                               builtins::Request::create(cx, requestInstance, input, headers_val));
+      options.request_headers = host_api::HttpReq(builtins::Request::request_handle(request));
+    }
+  }
+  return options;
+}
+
+// The JavaScript TransactionUpdateOptions parameter we are parsing should have the below interface:
+// interface TransactionUpdateOptions {
+//   maxAge: number,
+//   vary?: Array<string>,
+//   initialAge?: number,
+//   staleWhileRevalidate?: number,
+//   surrogateKeys?: Array<string>,
+//   length?: number,
+//   userMetadata?: ArrayBufferView | ArrayBuffer | URLSearchParams | string,
+// }
+JS::Result<host_api::CacheWriteOptions> parseTransactionUpdateOptions(JSContext *cx,
+                                                                      JS::HandleValue options_val) {
+  host_api::CacheWriteOptions options;
+  if (!options_val.isObject()) {
+    JS_ReportErrorASCII(cx, "options argument must be an object");
+    return JS::Result<host_api::CacheWriteOptions>(JS::Error());
+  }
+  JS::RootedObject options_obj(cx, &options_val.toObject());
+
+  JS::RootedValue maxAge_val(cx);
+  if (!JS_GetProperty(cx, options_obj, "maxAge", &maxAge_val)) {
+    return JS::Result<host_api::CacheWriteOptions>(JS::Error());
+  }
+  // maxAge property is required
+  if (maxAge_val.isUndefined()) {
+    JS_ReportErrorASCII(cx, "maxAge is required");
```

Comment:
What do you think about merging these two? We'd overwrite the error about field lookup with the `maxAge is required` message, but I think that might be more useful?

Reply:
That's a really nice improvement for the end-user 🙌